docs/design/jet/025-jet-job-member-pruning.md

```diff
@@ -0,0 +1,141 @@
+opaque to Jet. If `p1` is null for a member, `p2` has no input. But `p2` is still allowed to generate data, 
+so we can’t eliminate it. For example, if `p2` outputs the sum of input items, it still might want to emit `0`, 
+if there’s no input. 
+
+To support this behaviour, we need to extend the API. One option is add a method to PMS:
+```java
+interface ProcessorMetaSupplier extends IdentifiedDataSerializable {
+    // ...
+    default boolean doesWorkWithoutInput() {
+        return true;
+    }
+    // ...
+}
+```
+The default return value is `true` for backwards compatibility. This means that all current and new processors that 
+do no work without an input will have to override this method.
+
+Second, we need to support that solution in `ExecutionPlanBuilder` (execution plan creation phase). 
+Previous `ProcessorMetaSupplier#get` contract was a reason for strict assumptions in execution plan creation algorithm, 
+but that has changed. Moreover, we need to change algorithm in `ExecutionPlanBuilder` to accept `nulls`, and this is way
+trickier part to solve.
+
+There are two main approaches were considered, to shorten their meaning we mark them as 'heavyweight' and 'lightweight'.
+Long story short, the difference between these two approaches is an exact stage when we understand that some member
+may be pruned from job execution: 'heavyweight' approach tries to be more generic and define it for any accept DAG, when
+'lightweight' approach focus on hinting `ExecutionPlanBuilder` with additional metainformation from DAG and JobConfig 
+constructed by SQL optimizer.
+
+More details are available in chapters below.
+
+#### Heavyweight approach
+
+The core idea of that approach is that  `ExecutionPlanBuilder` would have detailed analysis of received DAG before 
+execution plan creation and then, possibly, trying to optimize DAG and only then apply member pruning based. Here, the
+requirements to prune member are : 
+- each vertex in the DAG may work without an input;
+- DAG does not contain `distributed-broadcast` edges; 
+- even if DAG contains `distributed-broadcast` edges, do they have any  ; 
+As a pros, using this approach generify member pruning usage for both SQL and Pipeline API. 
+As a cons, this approach brings high complexity in problem analysis, long development and big list of corner cases, which
+are discussed in chapter below.
+
+##### Issues/corner cases for heavyweight approach.
+Note : to be written. Maybe author will involve ASCII art for better visibility.
+
+1. Source -> Transform -> Aggregation -> Sink.
+Edges to Aggregation vertex (as a rule)  may be distributed.
+
+#### Lightweight approach
+
+Unlike a way described above, we can think that member pruning would be beneficial mostly for small jobs. 
+The overwhelming majority of such jobs are generated by SQL, and our team effort is to replace Predicate API, 
+where `PartitionPredicate` is available for collocated data querying. 
```

Comment:
is it `collocated`? For me "collocated"="partition aligned" means that I have 2 IMaps, and related records (eg. for the same customer) are on the same node. 

`PartitionPredicate` AFAIR only limits single IMap lookup/scanning to selected partitions.